File: packages/js/src/options.ts
import type { Diagnostic } from "./diagnostics/diagnostic.js"

export const markdownDialects = ["gfm"] as const
export const conversionProfiles = [
  "jira",
  "confluence",
  "portableMarkdown",
] as const

export type MarkdownDialect = (typeof markdownDialects)[number]
export type ConversionProfile = (typeof conversionProfiles)[number]

export type ConversionResult<T> = {
  value: T
  diagnostics: Diagnostic[]
}

export type ConversionOptions = {
  /**
   * Currently only GitHub Flavored Markdown is supported.
   */
  markdownDialect?: MarkdownDialect
  /**
   * Accepted for API and CLI parity. Profiles do not change Phase 1 conversion
   * behavior yet.
   */
  profile?: ConversionProfile
  /**
   * When true, ADF input is validated against the pinned ADF schema before
   * ADF-to-Markdown conversion.
   */
  validateAdf?: boolean
  /**
   * Accepted for future ADF normalization controls. Phase 1 converters already
   * emit normalized ADF where they can, so this is currently a no-op.
   */
  normalizeAdf?: boolean
}

export type ResolvedConversionOptions = Required<ConversionOptions>

const markdownDialectSet = new Set<string>(markdownDialects)
const conversionProfileSet = new Set<string>(conversionProfiles)

/**
 * Applies default conversion options and rejects unsupported dialect/profile
 * values before a converter starts work.
 */
export function resolveConversionOptions(
  options: ConversionOptions = {},
): ResolvedConversionOptions {
  const markdownDialect = options.markdownDialect ?? "gfm"
  if (!markdownDialectSet.has(markdownDialect)) {
    throw new Error(
      `Unsupported markdownDialect '${String(markdownDialect)}'. Supported markdownDialect is 'gfm'.`,
    )
  }

  const profile = options.profile ?? "portableMarkdown"
  if (!conversionProfileSet.has(profile)) {
    throw new Error(
      `Unsupported profile '${String(profile)}'. Supported profiles are jira, confluence, and portableMarkdown.`,
    )
  }

  return {
    markdownDialect,
    profile,
    validateAdf: options.validateAdf ?? true,
    normalizeAdf: options.normalizeAdf ?? true,
  }
}

File: packages/python/src/adfmd/options.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Generic, Mapping, TypeVar, cast

from .diagnostics.diagnostic import Diagnostic

T = TypeVar("T")

MARKDOWN_DIALECTS = ("gfm",)
CONVERSION_PROFILES = ("jira", "confluence", "portableMarkdown")


@dataclass(frozen=True)
class ConversionOptions:
    """Options shared by the JS and Python converters.

    Only GitHub Flavored Markdown is supported today. The named profiles are
    accepted for API and CLI parity, but do not change Phase 1 conversion
    behavior yet. ``normalize_adf`` is accepted for future normalization
    controls; Phase 1 converters already emit normalized ADF where they can.
    """

    markdown_dialect: str = "gfm"
    profile: str = "portableMarkdown"
    validate_adf: bool = True
    normalize_adf: bool = True

    def __post_init__(self) -> None:
        _validate_markdown_dialect(self.markdown_dialect)
        _validate_profile(self.profile)


@dataclass(frozen=True)
class ConversionResult(Generic[T]):
    value: T
    diagnostics: list[Diagnostic] = field(default_factory=list)


ConversionOptionsInput = ConversionOptions | Mapping[str, Any] | None


def resolve_conversion_options(options: ConversionOptionsInput = None) -> ConversionOptions:
    if options is None:
        return ConversionOptions()
    if isinstance(options, ConversionOptions):
        return options
    if isinstance(options, Mapping):
        markdown_dialect = cast(
            str,
            options.get("markdown_dialect", options.get("markdownDialect", "gfm")),
        )
        profile = cast(str, options.get("profile", "portableMarkdown"))
        validate_adf = bool(
            options.get("validate_adf", options.get("validateAdf", True))
        )
        normalize_adf = bool(
            options.get("normalize_adf", options.get("normalizeAdf", True))
        )
        return ConversionOptions(
            markdown_dialect=markdown_dialect,
            profile=profile,
            validate_adf=validate_adf,
            normalize_adf=normalize_adf,
        )


def _validate_markdown_dialect(value: str) -> None:
    if value not in MARKDOWN_DIALECTS:
        raise ValueError(
            f"Unsupported markdown_dialect '{value}'. Supported markdown_dialect is 'gfm'."
        )


def _validate_profile(value: str) -> None:
    if value not in CONVERSION_PROFILES:
        raise ValueError(
            "Unsupported profile "
            f"'{value}'. Supported profiles are jira, confluence, and portableMarkdown."
        )
